### PR TITLE
Print heredocs consistently with lineSuffix(newline+contents+ending) 

### PR DESCRIPTION
**Pull request by jbielick**
## Purpose

Instead of ad-hoc code in various places plucking out heredoc `beging`, `contents`, and `ending` in order to print it more intelligently in certain cases, this changes the way heredocs are printed altogether so they can simply be printed in any context and still work correctly. This prevents heredocs appearing in various parts of the syntax tree from _absorbing_ other lines of code because they break after the heredoc literal and before the heredoc contents because they were manually plucked out the AST and inserted in the Doc contextually. 

fixes prettier/plugin-ruby#624
fixes prettier/plugin-ruby#586
fixes prettier/plugin-ruby#316
fixes prettier/plugin-ruby#301

## Approach

I used `lineSuffix`.

The heredoc is 3 components, `beging` (the token), `contents` (the initial content and ending newline), and `ending` (the token ending).

```ruby
<<-TEXT
stuff stuff
stuff
TEXT
```

`beging`=`<<-TEXT`
`contents`=`stuff stuff\nstuff\n`
`ending`=`TEXT`

#### lineSuffix

[There exists a command in prettier's set of builders called `lineSufix`](https://github.com/prettier/prettier/blob/master/commands.md#linesuffix). It's listed as used for implementing trailing comments. The doc contents you wrap in `lineSuffix()` are appended to the end of the line during printing. This seemed appropriate for this use case because of the following facts about heredocs:

1. A heredoc literal beging can appear anywhere and can be treated like a string literal. things can be chained to it, it may need a trailing comma, etc).

Conclusion: When printing a heredoc, more code should be able to print after the heredoc literal before the next line.

2. The next line break after the heredoc beging **is part of the heredoc contents. period.**

Conclusion: A line break must _always_ appear, after which the contents printed then the ending. Code can continue on the next line. 

Conclusion: Code chained or in the same expression as the heredoc beging must fit on the same line as the heredoc beging or be broken and continue on the newline after the ending.

3. Using `lineSuffix` when printing heredocs allows us to guarantee an append to the end of that line where we can insert a literalline (non-negotiable line break), the contents, and the ending. Code continues printing after this ending newline.

Conclusion: `lineSuffix`'s behavior matches the behavior of heredocs in general and would allow a heredoc to be printed correctly in any context. The heredoc print would look like this:

```js
concat([
  beging,
  lineSuffix(concat([literalline, concat(parts), ending]))
]);
```

In words: since a heredoc beging appears "here", print the beging "here", but guarantee the following contents will appear at the end of the "here" line: `linebreak, contents, ending`. 

#### Result:

Thus, the end of line break and next few lines are guaranteed to be the heredoc contents (no absorbing other lines), the beging is the only doc portion printed in place so trailing commas and method chaining works fine. At this point, _all heredocs are printing themselves correctly anywhere they appear in the AST/Doc. The code used to manually slice up the heredoc and print it correctly in various places (hash.js, array.js, calls.js, utils.js) could now be removed and everything would print valid ruby code.

#### Change in formatting for heredocs in most places

However, now that a lineSuffix with a literalline is appended to the line with the heredoc, the `literalline` seems to break the `group`s and all parent `group`s, resulting in broken-up expressions wherever a heredoc appears. Here's an example:

**Before these changes:**

This expression would stay the same before and after formatting:

```ruby
call(1, <<-TEXT.squish)
  a sentence
TEXT
```

**After these changes:**

```ruby
call(
  1,
  <<-TEXT.squish
  a sentence
TEXT
)
```

I suspect because we're adding a literalline in the linesuffix of a heredoc printed doc, the result is that all parent groups are broken and can no longer be printed on a single line.

My opinion is that this is not undesirable, products consistent heredocs everywhere with more vertical space and potentially less confusion with those less familiar with heredocs. For example, two heredocs on the same line would now print differently:

Input:

```ruby
command 1, 2, 3, <<-HERE, <<-THERE
  here
HERE
  there
THERE
```

Becomes:

```ruby
command 1,
        2,
        3,
        <<-HERE,
  here
HERE
        <<-THERE
  there
THERE
```

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -6,7 +6,7 @@
   "scripts": {
     "lint": "eslint --cache .",
     "print": "prettier --plugin=.",
-    "test": "jest"
+    "test": "jest --runInBand"
```
Comment (dhughes) [inline package.json:9]: Is this something that's strictly necessary? If you're planning on submitting a PR upstream, then it might be undesirable to have unrelated/unnecessary changes. This just a thought/observation.

File: src/utils.js [hunk 2]
```diff
@@ -14,6 +8,27 @@ const containsAssignment = (node) =>
   ["assign", "massign"].includes(node.type) ||
   (node.type === "paren" && node.body[0].body.some(containsAssignment));
 
+const isHeredoc = (node) => {
```
Comment (dhughes) [inline src/utils.js:11]: I'm a bit unclear on the name of this method. It looks like this is returning an object, not a boolean as `is` would imply. Is this a convention used in plugin-ruby?

File: test/js/calls.test.js [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+      const content = ruby(`
+        foo = <<-TEXT.strip
+        bar
+        TEXT
+      `);
+
+      return expect(content).toMatchFormat();
+    });
+  });
+  test("with a call and indented", () => {
```
Comment (dhughes) [inline test/js/calls.test.js:45]: To be consistent, you might want a line break above this. 🤷

**Review by dhughes — APPROVED** [review]
I can't really profess to following everything in here, since you've dug into it much further than I have. But, I did look over it and I'm surprised by how much you were able to delete with this change. I left a couple questions, but nothing I'd think of as blocking.

Great work!!

File: package.json (hunk 1 above)
Comment (jbielick) [reply package.json]: preferential. it's now removed.

File: src/utils.js (hunk 2 above)
Comment (jbielick) [reply src/utils.js]: This code wasn't used, so I've removed it.
